services/QuillLMS/client/app/bundles/PremiumHub/components/usage_snapshots/filters.tsx

```diff
@@ -31,9 +34,110 @@ const Filters = ({
+
+  React.useEffect(() => {
+    initializePusher()
+  }, [pusherChannel])
+
+  React.useEffect(() => {
+    if (!pusherMessage) return
+
+    if (pusherMessage === getFilterHash({ key: FILTER_SCOPE_QUERY_KEY, id: diagnosticIdForStudentCount, withFilters: false })) {
+      getStudentCountData()
+    }
+    if (pusherMessage === getFilterHash({ key: FILTER_SCOPE_QUERY_KEY, id: diagnosticIdForStudentCount, withFilters: true })) {
+      getStudentCountData()
+    }
+  }, [pusherMessage])
+
+  React.useEffect(() => {
+    if (isGrowthDiagnosticReport && diagnosticIdForStudentCount && hasAdjustedFiltersFromDefault && !totalStudentCountForFilters && !totalStudentMatchesForFilters) {
+      getStudentCountData()
+    }
+  }, [diagnosticIdForStudentCount])
+
+  React.useEffect(() => {
+    if (totalStudentCountForFilters && totalStudentMatchesForFilters) {
+      setLoadingStudentCount(false)
+      setApplyFilterButtonClicked(true)
+    }
+  }, [totalStudentCountForFilters, totalStudentMatchesForFilters])
+
+  React.useEffect(() => {
+    setApplyFilterButtonClicked(false)
+    resetCounts()
+  }, [diagnosticIdForStudentCount])
+
+  function initializePusher() {
+    pusherChannel?.bind(PUSHER_EVENT_KEY, (body) => {
+      const { message, } = body
+
+      setPusherMessage(message)
+    });
+  };
+
+  function getFilterHash({ key, id, withFilters }) {
+    const filterTarget = [].concat(
+      key,
+      id,
+      selectedTimeframe.value,
+      withFilters ? selectedSchools.map(s => s.id) : null,
+      withFilters ? selectedGrades.map(g => g.value) : null,
+      withFilters ? selectedTeachers.map(t => t.id) : null,
+      withFilters ? selectedClassrooms.map(c => c.id) : null,
+    )
+    return hashPayload(filterTarget)
+  }
+
+  function resetCounts() {
+    setTotalStudentMatchesForFilters(null)
+    setTotalStudentCountForFilters(null)
+  }
+
+  function getStudentCountData() {
+    resetCounts()
+    setLoadingStudentCount(true)
+    getStudentCountDataForFilters(false)
+    getStudentCountDataForFilters(true)
+  }
+
+  function getStudentCountDataForFilters(withFilters) {
+    const searchParams = {
+      query: FILTER_SCOPE_QUERY_KEY,
+      timeframe: selectedTimeframe.value,
+      school_ids: withFilters ? selectedSchools.map(s => s.id) : null,
+      teacher_ids: withFilters ? selectedTeachers.map(t => t.id) : null,
+      classroom_ids: withFilters ? selectedClassrooms.map(c => c.id) : null,
+      grades: withFilters ? selectedGrades.map(g => g.value) : null,
+      diagnostic_id: diagnosticIdForStudentCount
+    }
+
+    requestPost('/admin_diagnostic_students/report', searchParams, (body) => {
+      if (!body.hasOwnProperty('results')) {
+        return
+      } else {
+        const { results } = body
+        const { count } = results
+        if (withFilters) {
+          setTotalStudentMatchesForFilters(count)
+        } else {
+          setTotalStudentCountForFilters(count)
+        }
+      }
+    })
+  }
```

Comment:
This feels like a lot of code to add to a shared component when it's all only being used for one instance of it. I think we should refactor some out into a shared hook to keep this cleaner -- [ChatGPT](https://chat.openai.com/share/82354a87-2c9a-4308-8252-d27a2558e03b) is a good place to start with this.